### PR TITLE
Fix/flows

### PR DESCRIPTION
**Pull request by nikhil-roy**
Added the following flows:

- Pagination: For sending a batch of locations to the server
- Buffering: For limiting the number of the locations saved in the Database
- Callbacks for events: For sending any events after syncing the locations to the server
- Configs: Added granular config
- Automatic Stop Service: Automatically Stopping the services after some duration

Feature Doc: https://paper.dropbox.com/doc/Consumer-GPS-SDK-First-Release-Flow-and-Features--As0QK1ILQSXlmYXDHRby7UDTAg-tE0T0yZY1pMnQeXiiWhGN

TICKET: https://app.clubhouse.io/shuttl/story/7504/gps-android-sdk

### CODE REVIEW

File: locationsyncservice/src/main/java/com/shuttl/location_pings/config/open_lib/LocationsHelper.kt [hunk 1]
```diff
@@ -15,22 +15,36 @@ import okhttp3.Interceptor
...
 object LocationsHelper {
 
+    var callback: LocationPingService.LocationPingServiceCallback? = null
+    private val serviceConnection by lazy {
+        object : ServiceConnection {
+            override fun onServiceDisconnected(name: ComponentName?) {
+            }
+
+            override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
+                (service as LocationPingService.CustomBinder).getService().setCallbackAndWork(callback)
```
Comment (betaCode95) [inline locationsyncservice/src/main/java/com/shuttl/location_pings/config/open_lib/LocationsHelper.kt:30]: use **as?** if we need 2 services to be binding, might cause casting exception

File: locationsyncservice/src/main/java/com/shuttl/location_pings/data/repo/LocationRepo.kt [hunk 2]
```diff
@@ -6,44 +6,56 @@ import com.shuttl.location_pings.config.components.LocationRetrofit
...
-    fun addLocation(location: GPSLocation) = GlobalScope.async(Dispatchers.IO) {
+    val TAG: String = javaClass.name
+
+    fun addLocation(location: GPSLocation, bufferSize: Int) = GlobalScope.async(Dispatchers.IO) {
+        val rowsCount = locationsDao?.getRowsCount() ?: 0
+        if (rowsCount >= bufferSize) {
+            locationsDao?.deleteOldestLocation()
+        }
         locationsDao?.addLocation(location)
+        printLocations()
```
Comment (betaCode95) [inline locationsyncservice/src/main/java/com/shuttl/location_pings/data/repo/LocationRepo.kt]: unnecessary logging

File: locationsyncservice/src/main/java/com/shuttl/location_pings/data/repo/LocationRepo.kt [hunk 3]
```diff
@@ -6,44 +6,56 @@ import com.shuttl.location_pings.config.components.LocationRetrofit
...
                     }
                 } catch (e: Exception) {
                     e.printStackTrace()
                 }
+
             }
+            locations
         }
+
+    fun printLocations() {
```
Comment (betaCode95) [inline locationsyncservice/src/main/java/com/shuttl/location_pings/data/repo/LocationRepo.kt]: don't think this logging is needed

File: locationsyncservice/src/main/java/com/shuttl/location_pings/service/LocationPingService.kt [hunk 4]
```diff
@@ -3,37 +3,59 @@ package com.shuttl.location_pings.service
...
+                            ?: "", configs.syncUrl ?: "",
+                        configs.userId ?: "",
+                        configs.bookingId ?: "",
+                        configs.batchSize
+                    )
+                val locations = a.await()
+                callback?.afterSyncLocations(locations)
+            }
+        } catch (e: Exception) {
+            Log.d("LocationsHelper", e.toString())
```
Comment (betaCode95) [inline locationsyncservice/src/main/java/com/shuttl/location_pings/service/LocationPingService.kt]: we should make this Log.e

File: locationsyncservice/src/main/java/com/shuttl/location_pings/service/LocationPingService.kt [hunk 5]
```diff
@@ -44,23 +66,37 @@ class LocationPingService : Service() {
...
             e.printStackTrace()
         }
     }
+
+    fun setCallbackAndWork(c: LocationPingServiceCallback?) {
+        callback = c
+        work()
+    }
+
+    interface LocationPingServiceCallback {
```
Comment (betaCode95) [inline locationsyncservice/src/main/java/com/shuttl/location_pings/service/LocationPingService.kt]: let's move this to different file

File: locationsyncservice/src/main/java/com/shuttl/location_pings/service/LocationSaveService.kt [hunk 6]
```diff
@@ -19,9 +20,21 @@ import com.shuttl.location_pings.data.repo.LocationRepo
 
 class LocationSaveService : Service() {
 
+    private val TAG: String = javaClass.name
     private val locManager by lazy { applicationContext.getSystemService(Context.LOCATION_SERVICE) as LocationManager }
     private var configs: LocationConfigs = LocationConfigs()
     private val repo by lazy { LocationRepo(LocationsDB.create(applicationContext)?.locationsDao()) }
+    private val timer by lazy {
+        object : CountDownTimer(configs.timeout.toLong(), 1000) {
+            override fun onTick(millisUntilFinished: Long) {
```
Comment (betaCode95) [inline locationsyncservice/src/main/java/com/shuttl/location_pings/service/LocationSaveService.kt:30]: can remove this override if not needed

File: locationsyncservice/src/main/java/com/shuttl/location_pings/data/repo/LocationRepo.kt [hunk 7]
```diff
@@ -6,44 +6,56 @@ import com.shuttl.location_pings.config.components.LocationRetrofit
...
+    suspend fun syncLocations(apiKey: String = "", url: String = "", userId: String = "", bookingId: String = "", batchSize: Int) = GlobalScope.async(Dispatchers.IO) {
+            val locations = locationsDao?.getLimitedLocations(batchSize)
+            if (locations?.isNotEmpty() == true) {
                 try {
                     if (TextUtils.isEmpty(url)) {
-                        Log.e("LocationsHelper", "No Url Found")
-                        return@launch
+                        Log.e(TAG, "No Url Found")
+                    } else {
+                        Log.i(TAG, "url: $url")
```
Comment (betaCode95) [inline locationsyncservice/src/main/java/com/shuttl/location_pings/data/repo/LocationRepo.kt]: unnecessary logging

File: locationsyncservice/src/main/java/com/shuttl/location_pings/service/LocationSaveService.kt (hunk 6 above)
Comment (nikhil-roy) [reply locationsyncservice/src/main/java/com/shuttl/location_pings/service/LocationSaveService.kt:30]: its abstract

**Review by betaCode95 — APPROVED** [review]

**Review by kshamashuttl — APPROVED** [review]

**Pull request merged**